packages/lib/survey/service.ts

```diff
@@ -623,37 +729,45 @@ export const getSyncSurveys = async (environmentId: string, person: TPerson): Pr
         }
       });
 
-      const attributeClasses = await getAttributeClasses(environmentId);
+      const personActions = await getActionsByPersonId(person.id);
+      const personActionClassIds = Array.from(
+        new Set(personActions?.map((action) => action.actionClass?.id ?? ""))
+      );
 
-      // filter surveys that meet the attributeFilters criteria
-      const potentialSurveysWithAttributes = surveys.filter((survey) => {
-        const attributeFilters = survey.attributeFilters;
-        if (attributeFilters.length === 0) {
-          return true;
-        }
-        // check if meets all attribute filters criterias
-        return attributeFilters.every((attributeFilter) => {
-          const attributeClassName = attributeClasses.find(
-            (attributeClass) => attributeClass.id === attributeFilter.attributeClassId
-          )?.name;
-          if (!attributeClassName) {
-            throw Error("Invalid attribute filter class");
-          }
-          const personAttributeValue = person.attributes[attributeClassName];
-          if (attributeFilter.condition === "equals") {
-            return personAttributeValue === attributeFilter.value;
-          } else if (attributeFilter.condition === "notEquals") {
-            return personAttributeValue !== attributeFilter.value;
-          } else {
-            throw Error("Invalid attribute filter condition");
+      if (!surveys.every((survey) => !survey.userSegment?.filters?.length)) {
+        const surveyPromises = surveys.map(async (survey) => {
+          const { userSegment } = survey;
+
+          if (userSegment) {
+            const result = await evaluateSegment(
+              {
+                attributes: person.attributes,
+                actionIds: personActionClassIds,
+                deviceType,
+                environmentId,
+                personId: person.id,
+              },
+              userSegment.filters
+            );
+
+            if (result) {
+              return survey;
+            }
+
+            return null;
           }
         });
-      });
+
+        const promisesResult = await Promise.all(surveyPromises);
+        const filteredResult = promisesResult.filter((survey) => !!survey);
+
+        surveys = filteredResult as TSurvey[];
+      }
 
       const latestDisplay = displays[0];
 
       // filter surveys that meet the recontactDays criteria
-      surveys = potentialSurveysWithAttributes.filter((survey) => {
+      surveys = surveys.filter((survey) => {
```

Comment:
Could it make sense to change the order of the checks?
As the user segments checks are very computational extensive we might want to do them at the end. If a survey was just shown to the user or he still won't see a survey because of the recontactDays (what normally happens most of the time), we shouldn't compute all the user segments.
So I propose to do the recontact day checks first and only if we have surveys left we query the actions and do the userTargeting calculations.